kyo-core/shared/src/main/scala/kyo/Scope.scala

```diff
@@ -0,0 +1,126 @@
+package kyo
+
+import kyo.Result.Error
+import kyo.Result.Panic
+import kyo.Tag
+import kyo.kernel.ContextEffect
+
+opaque type Scope = Sync
+
+object Scope:
+
+    private val local = Local.init(Maybe.empty[Finalizer])
+
+    def ensure(v: => Any < (Async & Abort[Throwable]))(using Frame): Unit < Scope =
+        ensure(_ => v)
+
+    def ensure(callback: Maybe[Error[Any]] => Any < (Async & Abort[Throwable]))(using Frame): Unit < Scope =
+        withFinalizerUnsafe(f => Abort.get(f.ensure(callback)))
+
+    def acquireRelease[A, S](acquire: => A < S)(release: A => Any < (Async & Abort[Throwable]))(using Frame): A < (Scope & S) =
+        withFinalizerUnsafe { finalizer =>
+            acquire.ensureMap { resource =>
+                val result = finalizer.ensure(_ => release(resource))
+                Abort.get(result).andThen(resource)
+            }
+        }
+
+    def acquire[A <: java.io.Closeable, S](resource: => A < S)(using Frame): A < (Scope & S) =
+        acquireRelease(resource)(_.close())
+
+    def run[A, S](v: A < (Scope & S))(using Frame): A < (Async & S) =
+        run(1, true)(v)
+
+    def run[A, S](closeParallelism: Int, awaitClose: Boolean = true)(v: A < (Scope & S))(using Frame): A < (Async & S) =
+        Sync.Unsafe(run(new Finalizer.Unsafe, closeParallelism, awaitClose)(v))
+
+    def runFinalizer[A, S](closeParallelism: Int, awaitClose: Boolean = true)(v: Finalizer => A < (Scope & S))(
+        using Frame
+    ): A < (Async & S) =
+        Sync.withLocal(local) { parent =>
+            import AllowUnsafe.embrace.danger
+            val child = new Finalizer.Unsafe
+            parent.foreach(p => discard(p.ensure(child.close(_, closeParallelism))))
+            run(child, closeParallelism, awaitClose)(v(child))
+        }
+
+    private def run[A, S](finalizer: Finalizer, closeParallelism: Int, awaitClose: Boolean)(v: A < (Scope & S))(
```

Comment:
`awaitClose` is a new feature and indicates if the execution should wait for the finalizer to close. If `false`, the closing happens in the background without backpressure. We could have separate apis returning a pending `Sync` instead of `Async` when it's `false` but it doesn't seem worth it.